### DISCUSSION TITLE
macOS on Apple silicon, Eclipse 2022-12 (4.26.0), C debugging

### DISCUSSION THREAD

**Discussion by DonWills** [eclipse-cdt/cdt#322] [Q&A]
macOS Venture 13.2.1, Mac Mini M2 Pro, Eclipse 2022-12 (4.26.0), CDT 11.0.0.202211062329
I can compile and run a plain old C program, but I can't debug it.  It appears that GDB has not been ported to Apple silicon.
So I installed the experimental Eclipse LLDB support.  I was able to change the debug environment to LLDB.  When I start that, a Debug dialog pops up that says LLDB-MI failed: unknown reason.
A post someplace else says that LLDB-MI is no more, that there is some new architecture to run LLDB.  (LLDB is installed - I can run it from the command line.)

I'm stuck.  I first just want to debug C programs on my new MacMini (it's really fast BTW).  Down the road I'd like to cross compile and debug remotely using Segger J-Link to a small processor board.

Do I have a prayer to accomplish any of this with Eclipse CDT?  (If the answer is no, then changing to CLion may be a solution, but I much prefer Eclipse for a variety of reasons.  And no, I won't change desktop computer/operating systems.)

**Comment by jonahgraham** [comment]
At the moment there is no real support for host debugging on macOS for the reasons you have identified. Debugging remote targets on macOS works well and it a main supported flow. The maintainer of the [Eclipse Embedded CDT](https://eclipse-embed-cdt.github.io/) project uses macOS. Please let us know if there is something in the cross compiling + debugging that doesn't work.

**Comment by DonWills** [comment]
I'm back again.  I really dislike CLion and VSCode.  I would really like to use CDT for plain old macOS development - no embedded - no remote debugging.  This issue hasn't changed in 15 months.  CDT already supports clang via the gcc->clang link on Apple silicon, so the only missing part is gdb.  It appears that the folks at GNU will not redesign gdb to use the macOS W^X switcher API - so gdb is dead for macOS as the Intel architecture and Rosetta evaporate into macOS history.
For the future health of Eclipse CDT (not just on macOS), it is imperative that a pure LLVM toolchain be an option for CDT.  JetBrains has done it with CLion, so I know it can be done.  Is there anyone with the expertise available to make those changes to the CDT debugger?  I am willing to pay for that development.  Please advise me on how this can be facilitated.  TIA.

**Reply by jonahgraham** [reply]
@DonWills can you email me privately (jonah at https://kichwacoders.com/) I can try to put you together with people that may be able to facilitate this.

**Comment by DonWills** [comment]
Thank you Jonah.  One of the people who you put me in contact with helped me figure out how it can be made to work.  Essentially you have to install the Experimental LLDB debugger from the main Eclipse download site.  Then install the XCode command line tools to get lldb.  Then install VSCode to get the lldb-mi executable file (hidden away in the directory structure).  Make both lldb and lldb-mi available for execution (set path, chmod).  Change the Default Launch configuration in Eclipse to use LLDB instead of GDB.  Launch your C program with debug and IT WORKS!

Suggestion:  These software parts should be packaged together in the standard macOS Eclipse distribution for C/C++ so plain old C programmers (who are just Eclipse users) can debug their plain old C programs on macOS on current hardware.

**Reply by DonWills** [reply]
I spoke too soon.  I just now got back to testing CDT native on my M2 Pro Mac Mini.  Breakpoints don't work at all (using the lldb-mi borrowed from VSCode).  Interestingly, single step does work!  This is not good for the future of CDT as a viable C development environment.

**Comment by jld01** [comment]
@DonWills, I also observed a problem with breakpoints when experimenting with LLDB 17.0.6 on a Windows host. If you are compiling using a recent version of GCC, try forcing DWARF v4 debug info by adding `-gdwarf-4` to your compiler command line:

![image](https://github.com/eclipse-cdt/cdt/assets/45816183/366a8661-9f19-42bd-8661-ebd583563d5c)

Does that help?

**Reply by DonWills** [reply]
No joy.  The compiler accepted the parameter, but breakpoints are still ignored.

**Comment by vfelsner** [comment]
Hi, same issue on my end. Was about to buy CLion for this but still prefer Eclipse. Is there a way how we can bump a clean LLDB integration in the development roadmap?

**Reply by jld01** [reply]
@vfelsner you should find improved integration with LLDB in [CDT 12.2.0 M3](https://download.eclipse.org/tools/cdt/builds/12.2/cdt-12.2.0-m3/) and later builds.

**Comment by jld01** [comment]
Can you look at the GDB trace and confirm that breakpoints appear to be set correctly? Ref: https://github.com/eclipse-cdt/cdt/blob/main/FAQ/README.md#ive-been-asked-for-gdb-traces-where-can-i-find-them

**Comment by jld01** [comment]
CDT support for native C/C++ application development under macOS on Apple silicon is now significantly improved. It is necessary to install a very recent build of the `lldb-mi` debugging tool. Setup details are described in the macOS section of the [Before you Begin](https://github.com/eclipse-cdt/cdt/blob/main/doc/org.eclipse.cdt.doc.user/src/concepts/cdt_c_before_you_begin.adoc#macos) documentation page.
